### ISSUE TITLE
Conversation Tracker for Strategy Logic, Dependents/Dependencies & Inline Script Execution Handling

### ISSUE THREAD

**Issue by 10upsimon** [10up/wordpress-develop#63]
In a bid to make sense of where we are at given the many concurrent Trac, PR, internal and meeting comms relating to async, inline script handling etc, I've taken a look at our current code in the PR and done my best to summarize what I believe highlights the current business rules in place.

I've used the legend below to mark what I believe to be approved and agreed upon logic, what I feel are areas that are currently in discussion and areas we know to be under construction/work in progress (mostly related to the async/deferred inline script execution when in the after position). I'm happy to (and planning to) make an issue in the 10up forked repo that we can reference for ongoing comms/decisions around this.

✅ = Assumed to be working as expected/discussed/agreed
⚠️  = Needs discussion points, or is in discussion
🚧  = Currently being iterated on in core PR or associated PR's

#### Script Dependencies / Dependents

- ✅ A blocking script, can have one or more blocking dependencies
- ✅ A blocking script, can have one or more deferred dependents
- ✅ A script is registered as deferred, but whom has dependents with blocking and/or deferred dependencies, remains deferred
- ✅ A script registered as deferred, but with dependents with async strategy, should have said dependents converted to defer
- ⚠️ An async script may have no dependencies (at present), or at least they won't remain async.
    - `We are proposing that async scripts may have dependencies and it is entirely up to the developer to know what they are doing and ensure that the async script handles itself correctly in terms of unknown timeline for dependency execution`
- ⚠️ Async scripts registered with dependencies of any strategy (at present), will become deferred
    - `See above proposed changes, this would negate this point.`
- ⚠️  Async scripts that have dependents, will be converted to a blocking strategy
    - `See above proposed changes, this would negate this point.`

#### Inline Script Handling

- ✅ Inline scripts attached to the before/after positions of blocking scripts, remain untouched and function as before
- ✅ Inline scripts attached to the before position of deferred scripts are acceptable and function the same as when attached to blocking scripts as the DOM order of scripts is preserved with deferred
- ✅   Non-standalone inline scripts attached to defer scripts in the after  position, are executed after the deferred script via a listener based script (iterated on and now merged to base PR)
- ✅ Standalone inline scripts attached to defer scripts in the after position are executed exactly the same as inline scripts attached to blocking scripts in the after position, i.e executed as and when they're parsed in the DOM
- ✅   Non-standalone inline scripts attached to async scripts in the after position, are executed after the deferred script via a listener based script (iterated on and now merged to base PR)
- ✅ Standalone inline scripts attached to async scripts in the after position are executed exactly the same as inline scripts attached to blocking scripts in the after position, i.e executed as and when they're parsed in the DOM
- ⚠️  Non-standalone inline scripts attached to the before  position of an async script, is not allowed at present.
    - `With @Weston Ruter new proposed means for handling inline scripts attached to defer or async main scripts, in any position, this should no longer be the case, and inline scripts in the after position (that are non standalone) should be able to be executed in a reliable order.`
-  ⚠️  Standalone inline scripts
    - `It is being discussed whether there is truly a need for this, or whether it over complicates the proposed API changes`

**Comment by joemcgill** [comment]
Thanks for laying out where we are with the remaining conversations, @10upsimon. I am generally in agreement with how you have things described here, with a couple of notable exceptions that I'm currently unsure about:

**Script Dependencies / Dependents**

> ✅ A script is registered as deferred, but whom has dependents with blocking and/or deferred dependencies, remains deferred

My understanding was that if a deferred script had a dependent that was blocking, the eligible loading strategy for that script would be converted to blocking in order to maintain execution order safety for the dependents.

> ⚠️ Async scripts that have dependents, will be converted to a blocking strategy

If an `async` script has any blocking dependents, it should become blocking. If it has no blocking dependents, but _does_ have dependents with a `defer` strategy, it should become `defer` as well. Both of these accommodations are meant specifically as a back-compat guard that makes it safer for scripts that are currently blocking to be converted to an `async` loading strategy without risking breakage of any dependent scripts.

**Inline Script Handling**

> 🚧 Non-standalone inline scripts attached to defer scripts in the after position, are executed after the deferred script via a listener based script (currently being iterated on)
> 🚧 Non-standalone inline scripts attached to async scripts in the after position, are executed after the deferred script via a listener based script (currently being iterated on)

For both these cases, we need to include unit tests to ensure that both the inline script is executed after the main async or deferred script _and also_ that we maintain loading order for any dependents of the main script such that the main script executes, then the inline script, and finally the dependents. This is to ensure backwards compatibility for any script that depends on both the initial async/deferred script and the inline script executing before it does.

> ⚠️ Non-standalone inline scripts attached to the before position of an async script, is not allowed at present.

Not sure I'm understanding this point. I didn't think we needed to adjust anything with `before` inline scripts. What question still needs to be resolved here?

> Standalone inline scripts (is there a need)

[I asked in the original PR](https://github.com/WordPress/wordpress-develop/pull/4391#discussion_r1198327475) if we should consider removing this optional param from the API as long as we're keeping the backwards compatibility intact.

One of the trickiest considerations is that (for backwards compatibility) we MUST assume that if an inline script is attached to a handle that has dependencies, those dependencies need to be loaded with a blocking strategy because the inline script may also depend on the execution of those dependencies. Adding the `standalone` parameter to the API allows developers to indicate that this is not the case, which would allow those dependencies to keep an async or deferred loading strategy, if applicable.

If we're ok with the limitations this puts on deferring some scripts, or if we were handling these cases by managing the execution of these before inline scripts by making sure they run after all deferred dependencies, then I think we could remove this complexity.

@felixarntz and @westonruter, would appreciate your thoughts on the above.

**Comment by westonruter** [comment]
@10upsimon:

> **Script Dependencies / Dependents**

> * ✅ A blocking script, can have one or more deferred dependencies

I believe you mean _dependents_ here, right? A blocking script cannot have an async/defer dependency, since the order would not be preserved.

But yes, a blocking script can have any number of async/defer dependents. In other words, an async/defer script can have blocking dependencies, no problem. An async script can also have async dependencies.

> ⚠️ An async script may have no dependencies (at present), or at least they won't remain async.
> * `We are proposing that async scripts may have dependencies and it is entirely up to the developer to know what they are doing and ensure that the async script handles itself correctly in terms of unknown timeline for dependency execution`

Correct. An async script may have dependencies of any strategy.

If a defer script has an async dependency, the async dependency should be converted to defer to ensure the order is preserved.

**Comment by westonruter** [comment]
@10upsimon:

> * 🚧  Non-standalone inline scripts attached to defer scripts in the after  position, are executed after the deferred script via a listener based script (currently being iterated on)
> * 🚧 Non-standalone inline scripts attached to async scripts in the after position, are executed after the deferred script via a listener based script (currently being iterated on)

I believe these are now ✅ 

> ⚠️ Non-standalone inline scripts attached to the before position of an async script, is not allowed at present.

Do you mean rather the `after` position? I won't see why a `before` inline script would need to have any impact on an `async` script.

> ⚠️ Standalone inline scripts

If we can't clearly demonstrate the use case (with examples from the ecosystem) and clearly communicate why these are needed, I do think we should consider removing them.

**Comment by westonruter** [comment]
@joemcgill:

> One of the trickiest considerations is that (for backwards compatibility) we MUST assume that if an inline script is attached to a handle that has dependencies, those dependencies need to be loaded with a blocking strategy because the inline script may also depend on the execution of those dependencies. Adding the `standalone` parameter to the API allows developers to indicate that this is not the case, which would allow those dependencies to keep an async or deferred loading strategy, if applicable.

I'm trying to understand the scenario here better.

Let's say we have two scripts--A and B--both of which are blocking. If they are both made defer, the issue at hand is that the `before` script for B could try accessing something from A, is that right?

1. A defer
2. B inline script: not deferred and accesses A
3. B defer

Step 2 would cause an error because A is not evaluated yet.

So is the current design intended to prevent A and B from being deferred unless the inline script for B is explicitly marked as `standalone`?

**Comment by 10upsimon** [comment]
> Let's say we have two scripts--A and B--both of which are blocking. If they are both made defer, the issue at hand is that the `before` script for B could try accessing something from A, is that right?
> 
> 1. A defer
> 2. B inline script: not deferred and accesses A
> 3. B defer
> 
> Step 2 would cause an error because A is not evaluated yet.
> 
> So is the current design intended to prevent A and B from being deferred unless the inline script for B is explicitly marked as `standalone`?

@westonruter @joemcgill just to emphasise that how we're handling it at present already handles and negates the above outcome. If you look at point 2 under `"Defer if"` at https://github.com/10up/wordpress-develop/issues/13#issuecomment-1446019943 this was the final rule.

This is handled in the code at https://github.com/WordPress/wordpress-develop/pull/4391/files#diff-65a2f96e438d914ebca50c9eb6b80dca7ab942fbd2ae80491af9da9db0bc32ffR982

In summary of the above: 

"A script is only deferred if there is no blocking dependent script and there is no `non-standalone` `before` script to any of the dependents."

The moment we opt in the inline script as `standalone`, this negates the non deferral rule and the main script would then be loaded deferred.

This is also validated by the test name `test_non_standalone_before_inline_script_with_defer_main_script` at https://github.com/WordPress/wordpress-develop/pull/4391/files#diff-09a2e7959959d6a08ef07b1849e0178b853f7ef86f7885bbf9be2ed261145b24R389

**Comment by 10upsimon** [comment]
> > **Script Dependencies / Dependents**
> 
> > * ✅ A blocking script, can have one or more deferred dependencies
> 
> I believe you mean _dependents_ here, right? A blocking script cannot have an async/defer dependency, since the order would not be preserved.

Totally correct, thanks for catching that @westonruter I'm updating now.

**Comment by joemcgill** [comment]
> A script is only deferred if there is no blocking dependent script and there is no non-standalone before script to any of the dependents.

Thinking about this some more...if we applied a similar event listener strategy to `before` inline scripts that we are applying `after` inline scripts (except we convert the inline script after dependencies are loaded) if there are any deferred dependencies, could we remove the `$standalone` parameter entirely? 

I think there's a case to be made that that inline scripts should by definition be treated as non-standalone, otherwise it would be better to use a different API to print those scripts.

**Comment by westonruter** [comment]
> Thinking about this some more...if we applied a similar event listener strategy to `before` inline scripts that we are applying `after` inline scripts (except we convert the inline script after dependencies are loaded) if there are any deferred dependencies, could we remove the `$standalone` parameter entirely?
> 
> I think there's a case to be made that that inline scripts should by definition be treated as non-standalone, otherwise it would be better to use a different API to print those scripts.

Yes, I think this should be explored. So if we have the above example:

* A defer script
* A inline after script: accesses A
* B inline before script: accesses A
* B defer script
* B inline after script: accesses B

We'll need to adjust the logic so that inline before scripts are also deferred in the same way as after scripts are. However, it won't be deferred to run after its associated script runs, but rather after all of the associated script dependencies run. In this case here, `B inline before script` needs to wait for `A defer` to execute.

**Comment by westonruter** [comment]
I've started investigating/prototyping this.

**Comment by joemcgill** [comment]
Just want to consolidate a few related conversations happening here, in the [Trac ticket conversation](https://core.trac.wordpress.org/ticket/12009#comment:115) and in the feature branch PR comments.

There remains some doubt that the eventual core commit for this feature should include support for inline scripts attached to `async` or `defer` scripts. I'd like to give a few days for @westonruter to finish his iteration that would keep support intact, but simplify the API by eliminating the need for the `$standalone` parameter on inline scripts by controlling the loading order for inline scripts in both the "before" AND "after" positions.

Once that's in place, I'd like to see a PR that removes the logic that loads inline scripts on event listeners (i.e., the `type="text/template"` approach) and instead handles backwards compatibility concerns by forcing scripts with inline dependents to be loaded as blocking scripts, so that themes and plugins can safely adopt the new loading strategies without worrying about breaking dependent inline scripts.

We can always enhance an initial implementation to add back support for inline scripts attached to `async` and/or `defer` scripts in a follow-up task, or include that functionality in an optional Performance Lab module in order to better evaluate whether this should be supported directly in core.

**Comment by westonruter** [comment]
> I'd like to give a few days for @westonruter to finish his iteration that would keep support intact, but simplify the API by eliminating the need for the `$standalone` parameter on inline scripts by controlling the loading order for inline scripts in both the "before" AND "after" positions.

Please see https://github.com/10up/wordpress-develop/pull/67. This PR reverts the introduction of standalone inline scripts while implementing support for delayed `before` inline scripts.

**Comment by joemcgill** [comment]
In order to not conflate the conversation here about how to handle inline scripts with the separate effort to remove deferred inline script handling in an initial commit, I've opened #69 as a separate issue that can be focused just on the details about removing delayed inline script handling. We can keep this conversation focused on ongoing implementation details for the preferred approach, which would allow inline scripts to be added to `async` and `defer` handles while maintaining execution order.

**Comment by joemcgill** [comment]
@10upsimon and @westonruter, now that #67 is merged, can we summarize the final intended status of all the use cases listed in the comment description and then close this one out, or is there additional work to do here?

**Comment by westonruter** [comment]
I don't think there is additional work to be done here, no.

To summarize the current state (note "delayed" means it's either async or defer):

* A blocking script can have any number of blocking dependencies (same as always).
* A blocking script can have any number of delayed dependents. (Delayed dependents can have blocking dependencies.)
* A delayed dependent can have any number of delayed dependencies of the same type. Defer dependents can have defer dependencies. Async dependents can have async dependencies.
* If a defer dependent has an async dependency, the async dependency is converted to defer in order to maintain the execution order.
* If a blocking dependent has delayed dependencies, all delayed dependencies are made blocking to maintain execution order.
* A dependency bundle (where `src` is `false`) is treated as a blocking dependency. Nevertheless, the bundle may consist of delayed dependents. 
  * This may not be fully fleshed out in the implementation, namely if a bundle has an async dependent but the bundle whole is a dependency of a `defer` dependent. We should make sure the `async` gets converted to `defer` here. See https://github.com/10up/wordpress-develop/issues/76.
* Scripts of any strategy may have inline before/after scripts. When a script is delayed, its after script will also be delayed (and before script will be delayed as well when relevant). The logic to handle the delayed execution of inline scripts is only added to the page when there are delayed inline scripts.

**Comment by joemcgill** [comment]
Thanks for the summary @westonruter. I'm going to close this conversation as complete and we can pick up any bugs discovered or considerations to change the summarized approach in separate issues.
